File: src-tauri/src/ai_runtime/providers/fal.rs
use crate::ai_runtime::errors::{AiResult, AiRuntimeError};
use crate::ai_runtime::polling::{cancelled_error, wait_interval_ms};
use crate::ai_runtime::providers::{ProviderContinuePollingInput, ProviderExecutionInput};
use crate::ai_runtime::task_registry;
use crate::ai_runtime::types::{GenerateStatus, ProviderExecutionResult};
use serde_json::{json, Value};

const FAL_SYNC_BASE_URL: &str = "https://fal.run";
const FAL_QUEUE_BASE_URL: &str = "https://queue.fal.run";

pub async fn execute(input: ProviderExecutionInput<'_>) -> AiResult<ProviderExecutionResult> {
    let sync_mode = input
        .body
        .get("sync_mode")
        .and_then(Value::as_bool)
        .unwrap_or(false);

    let clean_input = strip_sync_mode(input.body.clone());

    let submit_payload = if sync_mode {
        submit_sync(&input, &clean_input).await?
    } else {
        submit_async(&input, &clean_input).await?
    };

    if !sync_mode {
        let task_id = submit_payload
            .get("request_id")
            .and_then(Value::as_str)
            .filter(|value| !value.trim().is_empty())
            .map(ToString::to_string)
            .or_else(|| {
                submit_payload
                    .get("status_url")
                    .and_then(Value::as_str)
                    .filter(|value| !value.trim().is_empty())
                    .map(ToString::to_string)
            })
            .unwrap_or_else(|| input.request_id.to_string());

        return Ok(ProviderExecutionResult {
            status: GenerateStatus::Pending,
            url: String::new(),
            task_id: Some(task_id),
            metadata: submit_payload,
        });
    }

    let urls = extract_urls(&submit_payload);
    if urls.is_empty() {
        let request_id = submit_payload
            .get("request_id")
            .and_then(Value::as_str)
            .unwrap_or(input.request_id);
        return Err(AiRuntimeError::new(
            "empty_result",
            format!("Fal response has no media URL (request_id={})", request_id),
        ));
    }

    Ok(ProviderExecutionResult {
        status: GenerateStatus::Completed,
        url: urls.join("|||"),
        task_id: submit_payload
            .get("request_id")
            .and_then(Value::as_str)
            .map(ToString::to_string),
        metadata: submit_payload,
    })
}

pub async fn continue_polling(input: ProviderContinuePollingInput<'_>) -> AiResult<ProviderExecutionResult> {
    let request_id = input.task_id.to_string();
    let status_url = if input.task_id.starts_with("http://") || input.task_id.starts_with("https://") {
        input.task_id.to_string()
    } else {
        format!(
            "{}/{}/requests/{}/status",
            FAL_QUEUE_BASE_URL,
            input.route.trim_start_matches('/'),
            input.task_id
        )
    };
    let task_id = input.task_id.to_string();
    let final_payload = poll_by_status_url(input, &status_url).await?;
    let urls = extract_urls(&final_payload);
    if urls.is_empty() {
        return Err(AiRuntimeError::new(
            "empty_result",
            format!("Fal response has no media URL (request_id={})", request_id),
        ));
    }

    Ok(ProviderExecutionResult {
        status: GenerateStatus::Completed,
        url: urls.join("|||"),
        task_id: Some(task_id),
        metadata: final_payload,
    })
}

async fn submit_sync(input: &ProviderExecutionInput<'_>, clean_input: &Value) -> AiResult<Value> {
    let endpoint = normalize_endpoint(FAL_SYNC_BASE_URL, input.route);
    send_fal_request(input, &endpoint, clean_input).await
}

async fn submit_async(input: &ProviderExecutionInput<'_>, clean_input: &Value) -> AiResult<Value> {
    let endpoint = normalize_endpoint(FAL_QUEUE_BASE_URL, input.route);
    send_fal_request(input, &endpoint, clean_input).await
}

async fn send_fal_request(
    input: &ProviderExecutionInput<'_>,
    endpoint: &str,
    clean_input: &Value,
) -> AiResult<Value> {
    let payload = json!({ "input": clean_input });
    let response = input
        .client
        .post(endpoint)
        .header("Authorization", format!("Key {}", input.api_key))
        .header("Content-Type", "application/json")
        .json(&payload)
        .send()
        .await?;

    let status = response.status();
    let payload = response
        .json::<Value>()
        .await
        .map_err(|e| AiRuntimeError::new("invalid_json", e.to_string()))?;

    if !status.is_success() {
        return Err(AiRuntimeError::new(
            "provider_http_error",
            format!("Fal HTTP {}: {}", status, payload),
        ));
    }

    Ok(payload)
}

async fn poll_by_status_url(
    input: ProviderContinuePollingInput<'_>,
    status_url: &str,
) -> AiResult<Value> {
    let interval = input.polling.map(|p| p.interval).unwrap_or(3000);

    loop {
        if task_registry::is_cancelled(input.request_id) {
            return Err(cancelled_error(input.request_id));
        }

        wait_interval_ms(interval).await;

        let response = input
            .client
            .get(status_url)
            .header("Authorization", format!("Key {}", input.api_key))
            .send()
            .await?;

        let payload = response
            .json::<Value>()
            .await
            .map_err(|e| AiRuntimeError::new("invalid_json", e.to_string()))?;

        let state = payload
            .get("status")
            .and_then(Value::as_str)
            .unwrap_or("")
            .to_uppercase();

        if state == "COMPLETED" || state == "OK" {
            if let Some(response_url) = payload.get("response_url").and_then(Value::as_str) {
                let final_response = input
                    .client
                    .get(response_url)
                    .header("Authorization", format!("Key {}", input.api_key))
                    .send()
                    .await?;

                let mut response_json = final_response
                    .json::<Value>()
                    .await
                    .map_err(|e| AiRuntimeError::new("invalid_json", e.to_string()))?;
                if let Value::Object(map) = &mut response_json {
                    map.entry("request_id".to_string())
                        .or_insert_with(|| Value::String(input.task_id.to_string()));
                }
                return Ok(response_json);
            }
            let mut enriched = payload;
            if let Value::Object(map) = &mut enriched {
                map.entry("request_id".to_string())
                    .or_insert_with(|| Value::String(input.task_id.to_string()));
            }
            return Ok(enriched);
        }

        if state == "FAILED" || state == "ERROR" {
            return Err(AiRuntimeError::new("provider_task_failed", "Fal task failed"));
        }
    }
}

fn strip_sync_mode(value: Value) -> Value {
    if let Value::Object(mut map) = value {
        map.remove("sync_mode");
        return Value::Object(map);
    }
    value
}

fn extract_urls(payload: &Value) -> Vec<String> {
    let mut urls = Vec::new();
    collect_urls(payload, &mut urls);
    urls
}

fn collect_urls(value: &Value, target: &mut Vec<String>) {
    match value {
        Value::Object(map) => {
            for (key, item) in map {
                let lower = key.to_lowercase();
                if lower.contains("url") {
                    if let Some(url) = item.as_str() {
                        if url.starts_with("http://") || url.starts_with("https://") {
                            target.push(url.to_string());
                            continue;
                        }
                    }
                }
                collect_urls(item, target);
            }
        }
        Value::Array(items) => {
            for item in items {
                collect_urls(item, target);
            }
        }
        _ => {}
    }
}

fn normalize_endpoint(base: &str, route: &str) -> String {
    if route.starts_with("http://") || route.starts_with("https://") {
        return route.to_string();
    }
    if route.starts_with('/') {
        return format!("{}{}", base, route);
    }
    format!("{}/{}", base, route)
}

File: electron/main/services/ai-runtime/providers/fal.ts
import { AiRuntimeError } from '../errors'
import { pollUntilResult } from '../polling'
import type { JsonObject, JsonValue, ProviderContinuePollingInput, ProviderExecutionInput, ProviderExecutionResult } from '../types'
import { collectDeepUrls, normalizeEndpoint, readJsonResponse } from './helpers'
import { fetchProvider } from './provider-fetch'

const FAL_SYNC_BASE_URL = 'https://fal.run'
const FAL_QUEUE_BASE_URL = 'https://queue.fal.run'

export async function execute(input: ProviderExecutionInput): Promise<ProviderExecutionResult> {
  const syncMode = isJsonObject(input.body) && input.body.sync_mode === true
  const cleanInput = stripSyncMode(input.body)
  const payload = syncMode
    ? await submit(input, normalizeEndpoint(FAL_SYNC_BASE_URL, input.route), cleanInput)
    : await submit(input, normalizeEndpoint(FAL_QUEUE_BASE_URL, input.route), cleanInput)

  if (!syncMode) {
    const taskId = readString(payload, 'request_id') ?? readString(payload, 'status_url')
    if (!taskId) {
      throw new AiRuntimeError('empty_result', 'Fal queue response has neither request_id nor status_url')
    }
    return { status: 'pending', url: '', taskId, metadata: payload }
  }

  assertFalSucceeded(payload)

  const urls = extractUrls(payload)
  if (urls.length === 0) {
    throw new AiRuntimeError('empty_result', `Fal response has no media URL (request_id=${readString(payload, 'request_id') ?? input.requestId})`)
  }
  return { status: 'completed', url: urls.join('|||'), taskId: readString(payload, 'request_id'), metadata: payload }
}

export async function continuePolling(input: ProviderContinuePollingInput): Promise<ProviderExecutionResult> {
  const statusUrl = input.taskId.startsWith('http://') || input.taskId.startsWith('https://')
    ? input.taskId
    : `${FAL_QUEUE_BASE_URL}/${input.route.replace(/^\/+/, '')}/requests/${input.taskId}/status`
  const finalPayload = await pollByStatusUrl(input, statusUrl)
  const urls = extractUrls(finalPayload)
  if (urls.length === 0) {
    throw new AiRuntimeError('empty_result', `Fal response has no media URL (request_id=${input.taskId})`)
  }
  return { status: 'completed', url: urls.join('|||'), taskId: input.taskId, metadata: finalPayload }
}

async function submit(input: ProviderExecutionInput, endpoint: string, cleanInput: JsonValue): Promise<JsonValue> {
  const response = await fetchProvider('Fal', endpoint, {
    method: 'POST',
    headers: {
      Authorization: `Key ${input.apiKey}`,
      'Content-Type': 'application/json',
    },
    body: JSON.stringify(cleanInput),
    signal: input.signal,
  }, {
    retryPreconnectOnce: true,
  })
  return await readJsonResponse(response, 'Fal')
}

async function pollByStatusUrl(input: ProviderContinuePollingInput, statusUrl: string): Promise<JsonValue> {
  return await pollUntilResult(input, async () => {
    const response = await fetch(statusUrl, {
      headers: { Authorization: `Key ${input.apiKey}` },
      signal: input.signal,
    })
    const payload = await readJsonResponse(response, 'Fal')
    const state = (readString(payload, 'status') ?? '').toUpperCase()
    if (state === 'COMPLETED') {
      assertFalSucceeded(payload)
      const responseUrl = readString(payload, 'response_url')
      const resultUrl = responseUrl ?? statusUrl.replace(/\/status(?:\?.*)?$/, '')
      const finalResponse = await fetchProvider('Fal', resultUrl, {
        headers: { Authorization: `Key ${input.apiKey}` },
        signal: input.signal,
      }, {
        retryPreconnectOnce: true,
      })
      const result = await readJsonResponse(finalResponse, 'Fal')
      assertFalSucceeded(result)
      return enrichRequestId(result, input.taskId)
    }
    if (state === 'FAILED' || state === 'ERROR') {
      throw new AiRuntimeError('provider_task_failed', 'Fal task failed')
    }
    return undefined
  })
}

function assertFalSucceeded(payload: JsonValue): void {
  if (!isJsonObject(payload)) return
  const message = readString(payload, 'error')
    ?? readString(payload, 'detail')
    ?? readString(payload, 'message')
  const errorType = readString(payload, 'error_type')
  if (message || errorType) {
    throw new AiRuntimeError('provider_task_failed', message ?? errorType ?? 'Fal task failed')
  }
}

function stripSyncMode(value: JsonValue): JsonValue {
  if (!isJsonObject(value)) return value
  const next: JsonObject = { ...value }
  delete next.sync_mode
  return next
}

function extractUrls(payload: JsonValue): string[] {
  const urls: string[] = []
  collectDeepUrls(payload, urls)
  return urls
}

function enrichRequestId(value: JsonValue, requestId: string): JsonValue {
  return isJsonObject(value) ? { request_id: requestId, ...value } : value
}

function readString(value: JsonValue, key: string): string | undefined {
  return isJsonObject(value) && typeof value[key] === 'string' ? value[key] : undefined
}

function isJsonObject(value: JsonValue): value is JsonObject {
  return typeof value === 'object' && value !== null && !Array.isArray(value)
}
